nipype/interfaces/nilearn.py

```diff
@@ -0,0 +1,138 @@
+                                       'shared variance (e.g., for connectivity)')
+    include_global = traits.Bool(False, usedefault=True, mandatory=False,
+                                 desc='If True, include an extra column '
+                                 'labeled "global", with values calculated from the entire brain '
+                                 '(instead of just regions). Only has effect with 4D probability '
+                                 'maps.')
+    detrend = traits.Bool(False, usedefault=True, mandatory=False,
+                          desc='If True, perform detrending using nilearn.')
+
+class SignalExtractionOutputSpec(TraitedSpec):
+    out_file = File(exists=True, desc='tsv file containing the computed '
+                    'signals, with as many columns as there are labels and as '
+                    'many rows as there are timepoints in in_file, plus a '
+                    'header row with values from class_labels')
+
+class SignalExtraction(BaseInterface):
+    '''
+    Extracts signals over tissue classes or brain regions
+
+    >>> seinterface = SignalExtraction()
+    >>> seinterface.inputs.in_file = 'functional.nii'
+    >>> seinterface.inputs.label_files = 'segmentation0.nii.gz'
+    >>> seinterface.inputs.out_file = 'means.tsv'
+    >>> segments = ['CSF', 'gray', 'white']
+    >>> seinterface.inputs.class_labels = segments
+    >>> seinterface.inputs.detrend = True
+    >>> seinterface.inputs.include_global = True
+    '''
+    input_spec = SignalExtractionInputSpec
+    output_spec = SignalExtractionOutputSpec
+
+    def _run_interface(self, runtime):
+        maskers = self._process_inputs()
+
+        signals = []
+        for masker in maskers:
+            signals.append(masker.fit_transform(self.inputs.in_file))
+        region_signals = np.hstack(signals)
+
+        output = np.vstack((self.inputs.class_labels, region_signals.astype(str)))
+
+        # save output
+        np.savetxt(self.inputs.out_file, output, fmt=b'%s', delimiter='\t')
+        return runtime
+
+    def _process_inputs(self):
+        ''' validate and  process inputs into useful form.
+        Returns a list of nilearn maskers and the list of corresponding label names.'''
+        import nilearn.input_data as nl
+        import nilearn.image as nli
+
+        label_data = nli.concat_imgs(self.inputs.label_files)
+        maskers = []
+
+        # determine form of label files, choose appropriate nilearn masker
+        if np.amax(label_data.get_data()) > 1: # 3d label file
+            n_labels = np.amax(label_data.get_data())
+            maskers.append(nl.NiftiLabelsMasker(label_data))
+        else: # 4d labels
+            n_labels = label_data.get_data().shape[3]
+            if self.inputs.incl_shared_variance: # 4d labels, independent computation
+                for img in nli.iter_img(label_data):
+                    sortof_4d_img = nb.Nifti1Image(img.get_data()[:, :, :, np.newaxis], np.eye(4))
```

Comment:
Good to know about `img.affine`. `NiftiMapsMasker` complains if you input a 3-D nifti file, so that is my circuitous way of adding a 4th dimension (length 1). LMK if there is a native way to do it, I couldn't find one.